### PR TITLE
Separate out subtags into their own structs.

### LINKED ISSUE #55: Should `LanguageIdentifier::has_variant` have return value of `Result` type?
Hi @zbraniecki ! 

I am looking around the unic-langid API, specifically: https://docs.rs/unic-langid/0.8.0/unic_langid/struct.LanguageIdentifier.html#method.has_variant

It seems that the `Error`  branch is only because variant could be invalid.  Do you think it could be worth simplifying that API by factoring validation out of the `has` check?  The tradeoff is that we'd then need a `Variant` type (which is always a known-valid variant), but perhaps that is not entirely unreasonable.

I ask because it seems consistent with your recommendation that language ID parsing should be factored out of operations.  This is, of course, in line with the things going on at https://github.com/unicode-org/rust-discuss.

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #55.

The patch is not complete yet, but it comes to shape.
Unfortunately there's some performance penalty, which I think I'm ok swallowing. The tradeoff seems to be that we get a semantic wrapped around subtags, but some operations are slower.

I'm going to investigate performance more, I hope to erase the penalty for parser, but even if I fail, I'm willing to take it.

Current numbers are:

```
langid_canonicalize     time:   [698.57 ns 700.00 ns 702.05 ns]                                 
                        change: [+7.7270% +9.4705% +11.997%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

langid_canonicalize_from_bytes                                                                             
                        time:   [842.09 ns 843.41 ns 844.72 ns]
                        change: [+0.4536% +0.7349% +1.0119%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

Gnuplot not found, disabling plotting
     Running /projects/unic-locale/target/release/deps/langid-f1e84e1d0338254a
Gnuplot not found, disabling plotting
language_identifier_construct/from_str                                                                            
                        time:   [423.80 ns 424.52 ns 425.28 ns]
                        change: [+14.605% +14.964% +15.278%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) low mild
  1 (1.00%) high severe
language_identifier_construct/from_bytes                                                                            
                        time:   [418.97 ns 419.89 ns 420.96 ns]
                        change: [+17.675% +18.279% +19.137%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
language_identifier_construct/from_parts                                                                            
                        time:   [167.30 ns 167.46 ns 167.62 ns]
                        change: [-34.559% -34.398% -34.239%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

Gnuplot not found, disabling plotting
     Running /projects/unic-locale/target/release/deps/parser-b3f940d1f2b934f6
Gnuplot not found, disabling plotting
language_identifier_parser                                                                            
                        time:   [365.13 ns 366.00 ns 367.03 ns]
                        change: [+9.2740% +9.7830% +10.288%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

language_identifier_parser_casing                                                                            
                        time:   [364.51 ns 364.86 ns 365.23 ns]
                        change: [+6.5475% +6.8579% +7.1987%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
```

### CODE REVIEW

**Comment by zbraniecki** [comment]
I kept the `variants` private because we sort and dedup. At some point we may want to invest in `Variants` struct that would handle insert/remove/has/clear instead.

**Comment by zbraniecki** [comment]
I completed the port and cleaned up the parser getting it to be actually slightly faster than before:

```
langid_canonicalize     time:   [641.21 ns 643.18 ns 646.50 ns]                                 
                        change: [-3.0277% -2.7978% -2.5518%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

langid_canonicalize_from_bytes                                                                             
                        time:   [797.22 ns 798.39 ns 799.58 ns]
                        change: [-3.8163% -3.5905% -3.3481%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe

Gnuplot not found, disabling plotting
     Running target/release/deps/langid-f1e84e1d0338254a
Gnuplot not found, disabling plotting
language_identifier_construct/from_str                                                                            
                        time:   [335.91 ns 336.44 ns 337.03 ns]
                        change: [-6.6030% -6.2748% -5.9554%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low severe
  3 (3.00%) high mild
  3 (3.00%) high severe
language_identifier_construct/from_bytes                                                                            
                        time:   [330.94 ns 331.81 ns 333.04 ns]
                        change: [-6.4768% -6.0050% -5.6261%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  6 (6.00%) high severe
language_identifier_construct/from_parts                                                                             
                        time:   [39.045 ns 39.128 ns 39.220 ns]
                        change: [-84.545% -84.504% -84.463%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Gnuplot not found, disabling plotting
     Running target/release/deps/parser-b3f940d1f2b934f6
Gnuplot not found, disabling plotting
language_identifier_parser                                                                            
                        time:   [319.61 ns 320.20 ns 320.86 ns]
                        change: [-4.1813% -3.8953% -3.5983%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild

language_identifier_parser_casing                                                                            
                        time:   [320.19 ns 320.85 ns 321.61 ns]
                        change: [-1.0870% -0.7802% -0.4526%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
```

One change that I want to highlight is that I removed the helper methods that casted `LanguageIdentifier` field operations from `Locale` and instead I keep the `locale.id` exposed.

This is simpler and should be good enough and if we ever get to the point where we want to do something fancier, we can always add it.

**Comment by zbraniecki** [comment]
@Manishearth - this is basically separating out subtags for `unic_langid` as fields to move out parsing of the subtags out of helper methods.

**Comment by zbraniecki** [comment]
The `from_parts` is much faster because it basically takes over the role of `from_parts_unchecked` since it doesn't perform any operations.

The only reason I keep it as `unsafe` is because I need to assume the source is providing sorted variants. I think its fine for now and we'll likely end up with a custom sorted `Variants` struct.

File: unic-langid-impl/src/lib.rs [hunk 1]
```diff
@@ -108,51 +109,54 @@ impl LanguageIdentifier {
...
             script,
             region,
             variants,
-        })
+        }
+    }
+
+    /// # Safety
+    ///
+    /// This function accepts subtags expecting variants
```
Comment (Manishearth) [inline unic-langid-impl/src/lib.rs]: I don't think this needs to be `unsafe` anymore, just call it `unchecked`

**Comment by zbraniecki** [comment]
For the sake of providing easier feedback. This patch changes fallible setters/getters over strings of subtags like:

```rust
langid.set_language("foo").unwrap(); // because `foo` may not parse as language subtag
langid.has_variant("macos").unwrap(); // because `macos` may not parse as variant subtag
```

to:

```rust
pub struct LanguageIdentifier {
    pub language: subtags::Language,
    pub script: Option<subtags::Script>,
    pub region: Option<subtags::Region>,
    variants: Option<Box<[subtags::Variant]>>,
}
```

and then:

```rust
let lang: subtags::Language = "pl".parse().unwrap(); // because this can fail to parse
langid.langauge = Some(lang); // infallible and exposed field because you can't go wrong.

let variant: subtags::Variant = "macos".parse().unwrap(); // because this can fail
langid.has_variant(variant); // infallible!
```

**Comment by hagbard** [comment]
No "pub" for variants? Wouldn't that be a vector for variants (or is "Variant" really "Variants") ?

**Comment by zbraniecki** [comment]
> No "pub" for variants? Wouldn't that be a vector for variants (or is "Variant" really "Variants") ?

Potentially. I currently store variants sorted, so I need a setter to sort them, so that getter and `to_string` don't need to pay and returned value in getter is also sorted.

We could give up on that and allow for unsorted to be stored and sort on stringification, or we could come up with a custom `Variants` struct that will be a sorted `Vec`. I left it for later :)

**Comment by zbraniecki** [comment]
@hagbard - does it sound reasonable for you for now? `unic-locale` is not aiming for 1.0 anytime soon, so we can continue iterating.

**Comment by hagbard** [comment]
> > No "pub" for variants? Wouldn't that be a vector for variants (or is "Variant" really "Variants") ?
> 
> Potentially. I currently store variants sorted, so I need a setter to sort them, so that getter and `to_string` don't need to pay and returned value in getter is also sorted.
> 
> We could give up on that and allow for unsorted to be stored and sort on stringification, or we could come up with a custom `Variants` struct that will be a sorted `Vec`. I left it for later :)

Hmm, is sorting needed/preferred for comparison (I assume so). It feels weird to have a mix of pub fields and getters. I know Rust has an interesting relationship with mutability (nothing is inherently immutable at a struct level), but I personally (probably due to so much time with Java) think that this type of object, especially if it has invariants around things like sorting things or derived qualities like "is canonical", benefits a lot from being fully immutable.

My gut feeling is that we would be very unlikely to ever regret making it immutable by hiding fields and having only getters (cost of copy not big and not that common outside the class itself), but there's a non trivial chance we'd regret having public fields (e.g. no longer able to trivially cache some derived invariant). Furthermore if we did regret it not being mutable it would be easy to change that (but definitely not in the other direction).

**Comment by zbraniecki** [comment]
@hagbard thanks for the feedback! Since we're on the verge of importing this into ICU4X, I think we can continue the conv. in the PR that will get introduced soon!

I actually ended up also introducing `Variants` there to make all fields public and minimize the methods. I like the outcome, but I'll ask for your feedback as well there!

**Comment by hagbard** [comment]
It all depends on what we think about invariants and the cost of checking for equality or canonicalization etc. If we don't want to treat IDs with the same variants in a different order as "the same" then maybe its fine, or we stop using a plain list (which pretends iteration order is important) and either use a "self sorting" list, or just use a set (this is still problematic though because people *will* assume that iteration order is stable) .

Like all API design it's a tradeoff, so what important is to study usage and try to understand what people appear to want to do with this type. Then we can weight the options accordingly.

**Pull request merged**